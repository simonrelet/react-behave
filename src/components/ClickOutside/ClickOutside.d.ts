import * as React from 'react'

export interface RenderProps {
  ref: React.RefObject<HTMLElement>
}

export interface ClickOutsideProps {
  children(props: RenderProps): React.ReactNode
  onClickOutside(event: React.MouseEvent): void
  onEscape?(event: React.KeyboardEvent): void
}

export declare const ClickOutside: React.ComponentType<ClickOutsideProps>
