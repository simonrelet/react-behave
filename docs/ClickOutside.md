# ClickOutside

[create-ref]: https://reactjs.org/docs/react-api.html#reactcreateref

Notify for each click outside a component.

## Usage

```jsx
import React from 'react'
import { ClickOutside } from 'react-behave'

class App extends React.Component {
  handleClickOutside(mouseEvent) {
    console.log('click outside', mouseEvent)
  }

  handleEscape(keyEvent) {
    console.log('escape', keyEvent)
  }

  render() {
    return (
      <ClickOutside
        onClickOutside={this.handleClickOutside}
        onEscape={this.handleEscape}
      >
        {({ ref }) => <p ref={ref}>Don't click on me.</p>}
      </ClickOutside>
    )
  }
}
```

## Props

### `children`: `Function`

_Parameters_: `{ ref: Object|Function }`

Renders the component.
`ref` must be passed to the component in order to work.

### `onClickOutside`: `Function`

_Parameters_: `event: MouseEvent`

Called for each click outside the component.

### `onEscape`: `Function` (optional)

_Parameters_: `event: KeyboardEvent`

Called each time the Escape key is pressed.
