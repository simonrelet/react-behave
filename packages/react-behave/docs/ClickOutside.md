<!--
  THIS FILE WAS GENERATED!
  Don't make any changes in it, update src/components/ClickOutside/ClickOutside.js instead.
  If you still need to make changes in this file, remove this header so it won't be overridden.
-->

# ClickOutside

[create-ref]: https://reactjs.org/docs/react-api.html#reactcreateref

Notify for each click outside a component.

## Usage

```jsx
import React, { Component } from 'react';
import { ClickOutside } from 'react-behave';

class App extends Component {
  handleClickOutside(event) {
    console.log('click', event);
  }

  render() {
    return (
      <ClickOutside onClickOutside={this.handleClickOutside}>
        {ref => <p ref={ref}>Don't click on me.</p>}
      </ClickOutside>
    );
  }
}
```

## Props

### `children`: `Function`

_Parameters_: `ref: Object|Function`

Renders the component.
`ref` must be passed to the component in order to work.

### `onClickOutside`: `Function`

_Parameters_: `event: MouseEvent`

Called for each click outside the component.
