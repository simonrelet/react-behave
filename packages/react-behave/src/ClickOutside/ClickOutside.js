import PropTypes from 'prop-types';
import React, { Component } from 'react';
import EventListener from 'react-event-listener';

class ClickOutside extends Component {
  static propTypes = {
    onClickOutside: PropTypes.func.isRequired,
    render: PropTypes.func.isRequired,
  };

  elementRef = React.createRef();

  handleClickOutside = e => {
    const elementRef = this.elementRef.current;
    if (!elementRef || !elementRef.contains(e.target)) {
      this.props.onClickOutside(e);
    }
  };

  render() {
    return (
      <EventListener target="document" onClick={this.handleClickOutside}>
        {this.props.render({ ref: this.elementRef })}
      </EventListener>
    );
  }
}

export default ClickOutside;
