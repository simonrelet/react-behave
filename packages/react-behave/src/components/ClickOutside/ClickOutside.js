import PropTypes from 'prop-types'
import React, { Component } from 'react'
import EventListener from 'react-event-listener'

/**
 * [create-ref]: https://reactjs.org/docs/react-api.html#reactcreateref
 *
 * Notify for each click outside a component.
 *
 * ## Usage
 *
 * ```jsx
 * import React, { Component } from 'react';
 * import { ClickOutside } from 'react-behave';
 *
 * class App extends Component {
 *   handleClickOutside(event) {
 *     console.log('click', event);
 *   }
 *
 *   render() {
 *     return (
 *       <ClickOutside onClickOutside={this.handleClickOutside}>
 *         {ref => <p ref={ref}>Don't click on me.</p>}
 *       </ClickOutside>
 *     );
 *   }
 * }
 * ```
 */
class ClickOutside extends Component {
  static propTypes = {
    /**
     * _Parameters_: `ref: Object|Function`
     *
     * Renders the component.
     * `ref` must be passed to the component in order to work.
     */
    children: PropTypes.func.isRequired,

    /**
     * _Parameters_: `event: MouseEvent`
     *
     * Called for each click outside the component.
     */
    onClickOutside: PropTypes.func.isRequired,
  }

  elementRef = React.createRef()

  handleClickOutside = e => {
    const elementRef = this.elementRef.current
    if (elementRef && !elementRef.contains(e.target)) {
      this.props.onClickOutside(e)
    }
  }

  render() {
    return (
      <EventListener target="document" onClick={this.handleClickOutside}>
        {this.props.children(this.elementRef)}
      </EventListener>
    )
  }
}

export default ClickOutside
