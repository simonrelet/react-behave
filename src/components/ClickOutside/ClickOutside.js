import PropTypes from 'prop-types'
import React from 'react'

/**
 * [create-ref]: https://reactjs.org/docs/react-api.html#reactcreateref
 *
 * Notify for each click outside a component.
 *
 * ## Usage
 *
 * ```jsx
 * import React from 'react'
 * import { ClickOutside } from 'react-behave'
 *
 * class App extends React.Component {
 *   handleClickOutside(event) {
 *     console.log('click', event)
 *   }
 *
 *   render() {
 *     return (
 *       <ClickOutside onClickOutside={this.handleClickOutside}>
 *         {ref => <p ref={ref}>Don't click on me.</p>}
 *       </ClickOutside>
 *     )
 *   }
 * }
 * ```
 */
class ClickOutside extends React.Component {
  elementRef = React.createRef()

  componentDidMount() {
    document.addEventListener('mousedown', this.handleClickOutside)
  }

  componentWillUnmount() {
    document.removeEventListener('mousedown', this.handleClickOutside)
  }

  handleClickOutside = e => {
    const elementRef = this.elementRef.current
    if (elementRef && !elementRef.contains(e.target)) {
      this.props.onClickOutside(e)
    }
  }

  render() {
    return this.props.children(this.elementRef)
  }
}

ClickOutside.propTypes = {
  /**
   * _Parameters_: `ref: Object|Function`
   *
   * Renders the component.
   * `ref` must be passed to the component in order to work.
   */
  children: PropTypes.func.isRequired,

  /**
   * _Parameters_: `event: MouseEvent`
   *
   * Called for each click outside the component.
   */
  onClickOutside: PropTypes.func.isRequired,
}

export default ClickOutside
