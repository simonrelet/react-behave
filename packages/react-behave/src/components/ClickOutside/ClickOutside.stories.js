import { action } from '@storybook/addon-actions'
import centered from '@storybook/addon-centered'
import { storiesOf } from '@storybook/react'
import React from 'react'
import ClickOutside from './ClickOutside'

const stories = storiesOf('ClickOutside', module)

const style = {
  border: '1px solid #ccc',
  padding: '0.5rem 1rem',
}

stories.addDecorator(centered).add('Render function', () => (
  <ClickOutside onClickOutside={action('onClickOutside')}>
    {ref => (
      <p ref={ref} style={style}>
        Don't click on me.
      </p>
    )}
  </ClickOutside>
))
