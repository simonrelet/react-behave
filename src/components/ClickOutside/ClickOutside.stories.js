import centered from '@storybook/addon-centered'
import { storiesOf } from '@storybook/react'
import React from 'react'
import { ClickOutside } from './ClickOutside'

const stories = storiesOf('ClickOutside', module)

const style = {
  border: '1px solid #ccc',
  padding: '0.5rem 1rem',
}

stories.addDecorator(centered).add('Render function', () => (
  <ClickOutside onClickOutside={e => console.log('onClickOutside:', e)}>
    {({ ref }) => (
      <p ref={ref} style={style}>
        Don't click on me.
      </p>
    )}
  </ClickOutside>
))

stories.addDecorator(centered).add('Escape key', () => (
  <ClickOutside
    onClickOutside={e => console.log('onClickOutside:', e)}
    onEscape={e => console.log('onEscape:', e)}
  >
    {({ ref }) => (
      <p ref={ref} style={style}>
        Don't click on me or press Escape.
      </p>
    )}
  </ClickOutside>
))
