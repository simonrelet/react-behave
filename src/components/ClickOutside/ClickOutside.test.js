import React from 'react'
import { fireEvent, render } from 'react-testing-library'
import { ClickOutside } from './ClickOutside'

function createProps(props) {
  return {
    children: jest.fn(({ ref }) => <p ref={ref}>Hello</p>),
    onClickOutside: jest.fn(),
    ...props,
  }
}

describe('<ClickOutside />', () => {
  it('renders its content', () => {
    const props = createProps()
    const { container } = render(<ClickOutside {...props} />)
    expect(container.firstChild).toMatchSnapshot()
    expect(props.children).toHaveBeenCalledTimes(1)
  })

  it('notifies on a click outside', () => {
    const props = createProps()
    const { container } = render(<ClickOutside {...props} />)
    fireEvent.mouseDown(container)
    expect(props.onClickOutside).toHaveBeenCalledTimes(1)
  })

  it("doesn't notify on a click inside", () => {
    const props = createProps()
    const { container } = render(<ClickOutside {...props} />)
    fireEvent.mouseDown(container.firstChild)
    expect(props.onClickOutside).not.toHaveBeenCalled()
  })

  it("doesn't notify if there is no ref", () => {
    const props = createProps({ children: jest.fn(() => null) })
    const { container } = render(<ClickOutside {...props} />)
    fireEvent.mouseDown(container)
    expect(props.onClickOutside).not.toHaveBeenCalled()
  })
})
