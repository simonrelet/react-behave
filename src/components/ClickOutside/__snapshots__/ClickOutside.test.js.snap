// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClickOutside /> renders its content 1`] = `
<p>
  Hello
</p>
`;
